### PR TITLE
util: Fix a hash table collision bug

### PR DESCRIPTION
**Pull request by cvigue**
In util-hash.c there was some behavior that is unexpected and likely
incorrect. To see this behavior, create a hash table 32 entries wide
and use the default hash function. Then add a short string “abc”,
observe the string is stored properly. Now remove a string “iln”, and
observe string “abc” is no longer in the table.

This is because the hash function is not properly handling collisions in
some edge cases.

Includes new unit test:

- UT verifies that the hash function generates a collision for
  the selected test data. This must be true for the bug to be present.
  Then UT demonstrates the bug by adding two items to the hash table
  that collide, and then removing one of them 2x. The bug is that the
  other value is removed as well.

Bug #7828 --> https://redmine.openinfosecfoundation.org/issues/7828

Signed-off-by: Charlie Vigue <charlie.vigue@openvpn.com>

Make sure these boxes are checked accordingly before submitting your Pull Request -- thank you.

## Contribution style:
- [x] I have read the contributing guide lines at
   https://docs.suricata.io/en/latest/devguide/contributing/contribution-process.html

## Our Contribution agreements:
- [x] I have signed the Open Information Security Foundation contribution agreement at
   https://suricata.io/about/contribution-agreement/ (note: this is only required once)

## Changes (if applicable):
- [-] I have updated the User Guide (in [doc/userguide/](https://github.com/OISF/suricata/tree/304271e63a9e388412f25f0f94a1a0da4bf619d9/doc/userguide)) to reflect the changes made
- [-] I have updated the JSON schema (in [etc/schema.json](https://github.com/OISF/suricata/blob/304271e63a9e388412f25f0f94a1a0da4bf619d9/etc/schema.json)) to reflect all logging changes
      (including schema descriptions)
- [x] I have created a ticket at
      https://redmine.openinfosecfoundation.org/projects/suricata/issues

Link to ticket: https://redmine.openinfosecfoundation.org/issues/7828

Describe changes:

Added unit test to show issue
Changed remove to do a compare before removal
Consolidation of some common code via helpers

### Provide values to any of the below to override the defaults.

- To use a Suricata-Verify or Suricata-Update pull request,
  link to the pull request in the respective `_BRANCH` variable.
- Leave unused overrides blank or remove.

SV_REPO=
SV_BRANCH=
SU_REPO=
SU_BRANCH=

### CODE REVIEW

**Comment by inashivb** [comment]
@regit does this address the bug you reported in https://redmine.openinfosecfoundation.org/issues/7475?

**Comment by regit** [comment]
> @regit does this address the bug you reported in https://redmine.openinfosecfoundation.org/issues/7475?

yes, it looks like the same issue.

**Comment by cvigue** [comment]
There are some other oddities in the behavior of util-hash but the rest are generally probably best classified as undocumented functionality as far as I've noticed so far. For example, Add can silently add multiple identical items but the other functions don't really support dealing with the result of doing that.

**Comment by catenacyber** [comment]
How did you trigger this behavior in Suricata ? Where in the code do we try to remove an element from a hashtable when it was not added in it ?

**Comment by cvigue** [comment]
It's in the description and the unit test, discovered by code inspection.

**Comment by catenacyber** [comment]
> It's in the description and the unit test, discovered by code inspection.

I understand the unit test, but I do not know how this code is triggerable through an integration test like suricata-verify
And I would like to know the impact...

**Comment by cvigue** [comment]
There seem to be few calls to HashTableRemove currently and I'm unaware of any open defects that spring directly from this issue. The few places where HashTableRemove is called it doesn't seem like there is an explicit check for presence before removal. It might happen that the thing being removed will always be present as a coincidence or consequence of the surrounding code.  

It seems like counting on that forever being true isn't very robust though.

**Comment by inashivb** [comment]
> There seem to be few calls to HashTableRemove currently and I'm unaware of any open defects that spring directly from this issue. The few places where HashTableRemove is called it doesn't seem like there is an explicit check for presence before removal. It might happen that the thing being removed will always be present as a coincidence or consequence of the surrounding code.
> 
> It seems like counting on that forever being true isn't very robust though.

Thank you! So, low impact @catenacyber ?

**Comment by catenacyber** [comment]
> Thank you! So, low impact catenacyber ?

Likely low, but not fully assessed...

Could it be reached with datasets and unix commands to add and remove elements @inashivb ?

**Comment by inashivb** [comment]
> Could it be reached with datasets and unix commands to add and remove elements @inashivb ?

datasets use THash API (the thread safe one) not this one, I think. I'll try and see if I can trigger it with thresholding.

File: src/util-hash.c [hunk 1]
```diff
@@ -427,6 +426,39 @@ static int HashTableTestFull02 (void)
...
     return result;
 }
+
+static int HashTableTestCollisionBug(void)
+{
+    HashTable *ht = HashTableInit(32, HashTableGenericHash, NULL, NULL);
+    FAIL_IF_NULL(ht);
+
+    FAIL_IF_NOT(HashTableGenericHash(ht, (void *)"abc", 3) ==
+                HashTableGenericHash(ht, (void *)"iln", 3));
```
Comment (catenacyber) [inline src/util-hash.c:436]: How did you find strings that collide ?

File: src/util-hash.c [hunk 2]
```diff
@@ -138,44 +154,27 @@ int HashTableAdd(HashTable *ht, void *data, uint16_t datalen)
...
+ */
 int HashTableRemove(HashTable *ht, void *data, uint16_t datalen)
 {
     uint32_t hash = ht->Hash(ht, data, datalen);
 
-    if (ht->array[hash] == NULL) {
-        return -1;
-    }
-
-    if (ht->array[hash]->next == NULL) {
```
Comment (catenacyber) [inline src/util-hash.c:150]: Would not the fix be simply remove this `if` block ?

Or to add the condition like `if (ht->Compare(hashbucket->data,hashbucket->size,data,datalen) == 1)` to run it ?

**Comment by catenacyber** [comment]
`HashTableRemove` looks only used in 3 places
- detect-engine-threshold.c
- util-logopenfile.c
- util-mpm-hs.c

File: src/util-hash.c (hunk 1 above)
Comment (cvigue) [reply src/util-hash.c:436]: Just brute force

**Comment by inashivb** [comment]
> > Could it be reached with datasets and unix commands to add and remove elements @inashivb ?
> 
> datasets use THash API (the thread safe one) not this one, I think. I'll try and see if I can trigger it with thresholding.

This cannot be triggered with thresholding either.
The "lookup" key requires sid, gid and rev **AND** it is not allowed to have multiple threshold options in the same rule.

So, low impact?

File: src/util-hash.c (hunk 2 above)
Comment (cvigue) [reply src/util-hash.c:150]: That would work, however when I was reviewing the code it seemed more complex than it needed to be, and maybe that complication made it more error prone. So I simplified it down to one path while I was fixing the issue.

Comment (catenacyber) [reply src/util-hash.c:150]: I see your point, but then the diff is bigger and more complex...

So, I would prefer a simpler fix.
I guess we will talk about this with the team.

(By the way, if I did not mention it, the best fix would be to completely remove util-hash.c to only use the rust hashtables)

**Comment by catenacyber** [comment]
> This cannot be triggered with thresholding either. The "lookup" key requires sid, gid and rev **AND** it is not allowed to have multiple threshold options in the same rule.
> 
> So, low impact?

Or no impact at all ?

File: src/util-hash.c (hunk 2 above)
Comment (victorjulien) [reply src/util-hash.c:150]: A switch to rust would only be considered for 9+ though

Comment (catenacyber) [reply src/util-hash.c:150]: Yeah sure, I should have mentioned it

**Comment by victorjulien** [comment]
Merged in #13755, thanks!
